include/onnxruntime/core/framework/sparse_tensor.h

```diff
@@ -1,72 +1,317 @@
-#pragma once
-
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // Licensed under the MIT License.
 
+#pragma once
+
 #include "core/framework/data_types.h"
 #include "core/framework/tensor_shape.h"
 #include "core/framework/tensor.h"
 
 namespace onnxruntime {
 
+class IDataTransfer;
+class DataTransferManager;
+
+/**
+ * @brief This is a Sparse Format enumeration representing bitflags
+ * 
+ * 
+ */
+enum class SparseFormatFlags : uint32_t {
+  kUndefined = 0x0U,        // For completeness
+  kCoo = 0x1U,              // 1-D or 2-D indices
+  kCsrc = 0x1U << 1,        // Both CSR(C)
+  kBlockSparse = 0x1U << 2  // as in OpenAI
+};
+
+std::ostream& operator<<(std::ostream&, SparseFormatFlags);
+
+inline SparseFormatFlags operator|(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return static_cast<SparseFormatFlags>(static_cast<uint32_t>(flags) | static_cast<uint32_t>(flag));
+}
+
+inline SparseFormatFlags operator&(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return static_cast<SparseFormatFlags>(static_cast<uint32_t>(flags) & static_cast<uint32_t>(flag));
+}
+
+inline SparseFormatFlags TurnOff(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return static_cast<SparseFormatFlags>(static_cast<uint32_t>(flags) & ~static_cast<uint32_t>(flag));
+}
+
+inline bool IsSet(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return SparseFormatFlags::kUndefined != (flags & flag);
+}
+
+inline SparseFormatFlags Set(SparseFormatFlags flags, SparseFormatFlags flag) noexcept {
+  return flags | flag;
+}
+
+/**
+ * @brief This is a base class for virtual sparse format representation
+ * 
+ */
+class SparseRep {
+ protected:
+  SparseRep(MLDataType data_type, const TensorShape& values_shape, int64_t total_buffer_size, const AllocatorPtr& allocator);
+  SparseRep(MLDataType data_type, const TensorShape& values_shape, void* values, const OrtMemoryInfo& mem_info);
+
```

Comment:
nit: would `dense_shape` be clearer than `values_shape`?

Reply:
actually can you clarify the usage here as I'm not sure it's the dense_shape? what does values_shape represent and how does that relate to total_buffer_size? not clear to me as CSRC has 2 Tensor instances for indices.

and if `void* values` is pre-allocated why don't we need to know the size of that allocation?